modules/code-deployment/src/main/java/org/apache/ignite/internal/deployunit/DeploymentManagerImpl.java

```diff
@@ -0,0 +1,435 @@
+import org.apache.ignite.internal.deployunit.message.UndeployUnitRequestImpl;
+import org.apache.ignite.internal.deployunit.message.UndeployUnitResponse;
+import org.apache.ignite.internal.deployunit.message.UndeployUnitResponseImpl;
+import org.apache.ignite.internal.future.InFlightFutures;
+import org.apache.ignite.internal.logger.IgniteLogger;
+import org.apache.ignite.internal.logger.Loggers;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.metastorage.Entry;
+import org.apache.ignite.internal.metastorage.MetaStorageManager;
+import org.apache.ignite.internal.metastorage.dsl.Operation;
+import org.apache.ignite.internal.metastorage.dsl.Operations;
+import org.apache.ignite.internal.util.IgniteUtils;
+import org.apache.ignite.lang.ByteArray;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.ClusterService;
+
+//TODO: rework metastorage keys IGNITE-18870
+/**
+ * Deployment manager implementation.
+ */
+public class DeploymentManagerImpl implements IgniteDeployment, IgniteComponent {
+
+    private static final IgniteLogger LOG = Loggers.forClass(DeploymentManagerImpl.class);
+
+    private static final String TMP_SUFFIX = ".tmp";
+
+    private static final String DEPLOY_UNIT_PREFIX = "deploy-unit.";
+
+    private static final String UNITS_PREFIX = DEPLOY_UNIT_PREFIX + "units.";
+
+    /**
+     * Meta storage.
+     */
+    private final MetaStorageManager metaStorage;
+
+    /**
+     * Deployment configuration.
+     */
+    private final DeploymentConfiguration configuration;
+
+    /**
+     * Cluster management group manager.
+     */
+    private final ClusterManagementGroupManager cmgManager;
+
+    /**
+     * In flight futures tracker.
+     */
+    private final InFlightFutures inFlightFutures = new InFlightFutures();
+
+    /**
+     * Cluster service.
+     */
+    private final ClusterService clusterService;
+
+    /**
+     * Folder for units.
+     */
+    private Path unitsFolder;
+
+    /**
+     * Constructor.
+     *
+     * @param clusterService Cluster service.
+     * @param metaStorage Meta storage.
+     * @param workDir Node working directory.
+     * @param configuration Deployment configuration.
+     * @param cmgManager Cluster management group manager.
+     */
+    public DeploymentManagerImpl(ClusterService clusterService,
+            MetaStorageManager metaStorage,
+            Path workDir,
+            DeploymentConfiguration configuration,
+            ClusterManagementGroupManager cmgManager) {
+        this.clusterService = clusterService;
+        this.metaStorage = metaStorage;
+        this.configuration = configuration;
+        this.cmgManager = cmgManager;
+        unitsFolder = workDir;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> deployAsync(String id, Version version, DeploymentUnit deploymentUnit) {
+        if (id == null || id.isBlank() || version == null) {
```

Comment:
1. Let's check every argument separately and throw `IllegalArgumentException` directly. I don't think it should be wrapped into a future.

2. `deploymentUnit` check is missing.